extra/models/clip.py

```diff
@@ -222,9 +222,11 @@ def __init__(self, ret_layer_idx:Optional[int]=None):
     self.transformer = Closed.ClipTextModel(ret_layer_idx)
     self.input_key   = "txt"
 
-  def __call__(self, text:str) -> Union[Tensor,Tuple[Tensor,...]]:
-    tokens = Tensor(self.tokenizer.encode(text))
-    return self.transformer.text_model(tokens.reshape(1,-1))
+  def __call__(self, texts:Union[str,List[str],Tensor]) -> Union[Tensor,Tuple[Tensor,...]]:
```

Comment:
can `texts` be `Tensor` here?

same for other functions. I find mixing str and Tensor weird because we don't support string tensors

Reply:
This ends up being weird since it's the input to the `Embedder` which for CLIP is the raw text to get tokenized and then embedded, the CLIP tokenizer expects a string input to return a Tensor

The reason there is a `Union` between `str` and `Tensor` is that SDXL has multiple types of embedders, some acting on the text prompt and some acting on values like target image size, the typing was to make the system happy with both kinds of inputs, while each embedder starts with an assert now to make sure it got the right kind of input.

I understand your point, and in theory this could get converted, but it would require porting the entire tokenizer logic to tinygrad, which seems like a lot of effort for little gain.

On another note, the whole Embedder system was due to porting the structure in the reference implementation. The whole thing is causing issues in relationship to sharding, and I am currently looking into how I want this to get refactored. Everything is currently abstracted away and built around dependency injection, which is beneficial for researchers trying out multiple architecture configurations, but terrible for a direct implementation that is understandable. Making things more explicit should make it both more understandable and removed this kind of "dual input types".

Reply:
i see, asserting for now should be fine